### ISSUE TITLE
[nll] spurious outlives requirement in shred-0.7.0

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#53121]
The following snippet, extracted from the crate shred v0.7.0, fails to compile with NLL, **but only when compiled with `--crate-type lib`**: 

```rust
// NB You must compile with `--crate-type lib` to see the failure!

#![feature(nll)]

pub trait Accessor: Sized {
}

pub enum AccessorCow<'a, 'b, T>
where
    AccessorTy<'a, T>: 'b,
    T: System<'a> + ?Sized,
'a: 'b,
{
    /// A reference to an accessor.
    Ref(&'b AccessorTy<'a, T>),
    /// An owned accessor.
    Owned(AccessorTy<'a, T>),
}

type AccessorTy<'a, T> = <<T as System<'a>>::SystemData as DynamicSystemData<'a>>::Accessor;

pub trait System<'a> {
    type SystemData: DynamicSystemData<'a>;
}

pub trait SystemData<'a> {
}

pub trait DynamicSystemData<'a> {
    /// The accessor of the `SystemData`, which specifies the read and write dependencies and does
    /// the fetching.
    type Accessor: Accessor;
}

fn main() { }
```

yields

```
> rustc +nightly ~/tmp/reduce.rs -Zborrowck=mir -Ztwo-phase-borrows --crate-type lib

error[E0309]: the parameter type `T` may not live long enough
  --> /home/nmatsakis/tmp/reduce.rs:13:5
   |
13 |     Ref(&'b AccessorTy<'a, T>),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `T: 'b`...

error[E0309]: the parameter type `T` may not live long enough
  --> /home/nmatsakis/tmp/reduce.rs:15:5
   |
15 |     Owned(AccessorTy<'a, T>),
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `T: 'b`...
```

I'm not entirely sure what is going on here, but it seems very fishy, particularly since it is specific to `--crate-type lib`.

**Comment by nikomatsakis** [comment]
OK so I've dug into this a bit more. The NLL behavior here is buggy, I think, but I'm having trouble isolating it to a nicer test case because in general our behavior here is not *that* strong. I think that this struct *ought* to be well-formed as is -- requiring that `T: 'b` is sufficient but not **necessary**, the existing `where AccessorTy<'a, T>: 'b` ought to suffice.

I think that the problem is that the NLL code is hitting this "fallback" in the code that handles obligations like `AccessorTy<'0, T>: '1`:

https://github.com/rust-lang/rust/blob/cf84056149baf649f0c297b02afdde7db2187f46/src/librustc/infer/outlives/obligations.rs#L383-L410

In particular, if we encounter an obligation to prove `AccessorTy<'0, T>: '1`, we fallback to requiring that `'0: '1` and `T: '1`. As the comment describes, this case is quite tricky because if we *don't* add constraints to inference, then `'0` might be too small -- but if we *do*, we don't know which to add really. 

Anyway, since NLL uses region inference variables everywhere, it basically *always* hits this case, which makes us overly conservative in this example.  

I'm not sure yet the right fix. I suppose that in NLL we could avoid going down this path, but I think this will likely cause other failures.

**Comment by nikomatsakis** [comment]
I was thinking about this. I think maybe the best fix is that we should not be so eager to convert "outlives bounds" into type tests and constraints. We could hold off on that, do inference, and then do the type test conversion. This second round may of course add new constraints, and we can deal with that.

This will be a *bit* of a pain, though, since we currently compute the SCCs etc based on the previous set of obligations, and this would introduce new edges into the graph. Not sure how to deal with that. 

Gah, still not entirely sure what to do here.

**Comment by pnkfelix** [comment]
visited at meeting. assigning to niko to try to write up a plan of attack. (Also, we need to decide whether it is really an RC blocker or not)

**Comment by nikomatsakis** [comment]
Not an RC blocker: the migration strategy means this will be a warning anyway.

**Comment by nikomatsakis** [comment]
Minimized test showing the core problem:

```rust
trait MyTrait<'a> {
    type Output;
}

fn foo<'a, T>() -> &'a ()
where
    T: MyTrait<'a>, <T as MyTrait<'a>>::Output: 'a,
{
    bar::<T::Output>()
}

fn bar<'a, T>() -> &'a ()
where
    T: 'a,
{
    &()
}

fn main() {}
```

**Comment by nikomatsakis** [comment]
Fix in #54453
